programs/host/frontend.c

```diff
@@ -0,0 +1,943 @@
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  This file is part of mbed TLS (https://tls.mbed.org)
+ */
+
+#include "frontend-config.h"
+
+#include <stdint.h>
+#include <errno.h>
+
+#if defined(MBEDTLS_PLATFORM_C)
+#include "mbedtls/platform.h"
+#else
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#define mbedtls_calloc    calloc
+#define mbedtls_free      free
+#define mbedtls_fprintf   fprintf
+#include <unistd.h>
+#include <dirent.h>
+#endif
+
+FILE * fdbg = NULL;
+
+static int debug_verbose = 0;
+#define DBG( fmt, ... ) do {                                                \
+               fprintf( fdbg, "%s:%d:%s: " fmt "\n",                  \
+                        __FILE__, __LINE__, __FUNCTION__, ##__VA_ARGS__ );  \
+               fflush( fdbg ); \
+} while (0)
+
+#include "mbedtls/serialize.h"
+#include "mbedtls/net_sockets.h"
+
+/** State of the offloading frontend. */
+typedef enum {
+    MBEDTLS_SERIALIZE_STATUS_DEAD = 0, /**< The communication channel is broken */
+    MBEDTLS_SERIALIZE_STATUS_OK = 1, /**< All conditions nominal */
+    MBEDTLS_SERIALIZE_STATUS_OUT_OF_MEMORY = 2, /**< Out of memory for a function's parameters. Normal operation can resume after the next stack flush. */
+    MBEDTLS_SERIALIZE_STATUS_EXITED = 3, /**< An exit command has been received */
+} mbedtls_serialize_status_t;
+
+/** An input or output to a serialized function.
+ *
+ * Inputs are kept as a linked list, outputs are kept in an array. There's
+ * no deep reason for that, in retrospect an array would do fine for inputs. */
+struct mbedtls_serialize_item;
+typedef struct mbedtls_serialize_item {
+    struct mbedtls_serialize_item *next; /**< Next input */
+    size_t size; /**< Size of the following data in bytes */
+    /* followed by the actual data */
+} mbedtls_serialize_item_t;
+
+/** Specialization of mbedtls_serialize_item_t with enough room for
+ * data for a uint32. */
+typedef struct {
+    mbedtls_serialize_item_t meta;
+    unsigned char data[4];
+} mbedtls_serialize_uint32_t;
+
+
+static void *item_buffer( mbedtls_serialize_item_t *item )
+{
+    return( (unsigned char*) item + sizeof( *item ) );
+}
+
+static uint16_t item_int16( mbedtls_serialize_item_t *item )
+{
+    uint8_t *buffer = item_buffer( item );
+    return( buffer[0] << 8 | buffer[1] );
+}
+
+static uint32_t item_int32( mbedtls_serialize_item_t *item )
```

Comment:
Name these functions `item_uintXX` instead to avoid any false expectation regarding the type of their return value?